ping/manager.go

```diff
@@ -0,0 +1,215 @@
+		ipTargets:  conf.IPTargets,
+
+		pingInterval: conf.PingInterval,
+		pingTimout:   conf.PingTimout,
+	}
+}
+
+// Run starts the manager.
+// This will block until the context is canceled
+func (m *Manager) Run(ctx context.Context) error {
+	l := logr.FromContextOrDiscard(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-m.ticker.C:
+			err := m.schedulePingers(ctx)
+			// We don't want to fail if DNS disappears, just log and move on.
+			if err != nil {
+				l.Error(err, "Failed to schedule pingers")
+			}
+		}
+	}
+}
+func (m *Manager) schedulePingers(ctx context.Context) error {
+	l := logr.FromContextOrDiscard(ctx)
+	l.Info("Updating ping targets")
+	ips, err := m.lookUpTargets(ctx)
+	if err != nil {
+		return err
+	}
+	err = m.cleanupPingers(ctx, ips)
+	if err != nil {
+		return err
+	}
+	err = m.startPingers(ctx, ips)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Manager) cleanupPingers(ctx context.Context, ips []*net.IPAddr) error {
+	l := logr.FromContextOrDiscard(ctx)
+	for ip, pCtx := range m.pingCtxs {
+		if pCtx.Err() != nil {
+			l.Info("Removing failed pinger")
+			delete(m.pingCtxs, ip)
+		}
+		if !ipInList(ip, ips) {
+			l.Info("Stopping outdated pinger")
+			m.pingCtxs[ip].cancel()
+			delete(m.pingCtxs, ip)
+		}
+	}
+	return nil
+}
+
+func (m *Manager) startPingers(ctx context.Context, ips []*net.IPAddr) error {
+	l := logr.FromContextOrDiscard(ctx)
+	for i := range ips {
+		if _, ok := m.pingCtxs[ips[i].String()]; !ok {
+			l.Info("Starting pinger", "target", ips[i].String())
+			err := m.startPing(ctx, ips[i])
+			if err != nil {
+				l.Error(err, "Failed to start ping")
+			}
+		}
+	}
+	return nil
+}
+
+func ipInList(key string, ips []*net.IPAddr) bool {
+	for _, ip := range ips {
+		if ip.String() == key {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *Manager) lookUpTargets(ctx context.Context) ([]*net.IPAddr, error) {
+	ips := []*net.IPAddr{}
+	for _, t := range m.ipTargets {
+		ip, err := net.ResolveIPAddr("ip", t)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to resolve IP addr: %w", err)
+		}
+
+		ips = append(ips, ip)
+	}
+	for _, t := range m.dnsTargets {
+		res, err := m.lookupIP(t)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to resolve DNS addr: %w", err)
+		}
+		for _, ip := range res {
+			if ip.To4() == nil {
+				// Not an IPv4
+				continue
+			}
+			ips = append(ips, &net.IPAddr{
+				IP: ip,
+			})
+		}
+	}
+	return ips, nil
+}
+
+func (m *Manager) startPing(ctx context.Context, ip *net.IPAddr) error {
+	l := logr.FromContextOrDiscard(ctx)
+	l = l.WithName(ip.IP.String())
+
+	labels := prometheus.Labels{"source": m.source, "target": ip.IP.String()}
+	hist, err := m.hist.CurryWith(labels)
+	if err != nil {
+		return err
+	}
+	pinger, err := m.newPinger(ip, hist, m.pingInterval, m.pingTimout)
+	if err != nil {
+		return err
+	}
+	ctx = logr.NewContext(ctx, l)
+	ctx, cancel := context.WithCancel(ctx)
+	m.pingCtxs[ip.String()] = Context{
```

Comment:
There's now also also the comparable [`net/netip`](https://go.dev/doc/go1.18#netip). But I'm completely fine as is.

Reply:
That's actually a good input, I forgot that this existed and really aren't a fan of the current solution. However to do this properly I'd need to change quite a bit. I might do this cleanup after the initial release.